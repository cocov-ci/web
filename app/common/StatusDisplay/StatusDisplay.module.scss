@import "styles/variables.module";

.statusDisplay {
  .base {
    display: flex;
    flex-direction: row;
    padding: 0;
    font-size: 15px;

    .cell {
      flex: 1;
      display: flex;
      align-items: center;
      justify-content: space-between;

      .text {
        padding: 18px 25px;

        .dot {
          height: 12px;
          width: 12px;
        }

        .label {
          margin-left: 6px;
        }

        .message {
          margin-left: 4px;

          &.bold {
            font-variation-settings: "wght" 600;
          }
        }
      }

      .details {
        margin: 4px 8px;

        &.help {
          margin-right: 12px;
        }
      }

      &:first-of-type {
        border-right: 1px solid $colorGray;
      }
    }
  }
}
