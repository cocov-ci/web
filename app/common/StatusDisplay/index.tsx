import classNames from 'classnames'
import React from 'react'

import AccessoryButton, { HelpIcon } from 'app/common/AccessoryButton'
import Box from 'app/common/Box'
import StatusDot, { StatusDotColor } from 'app/common/StatusDot'
import Url from 'types/Url'

import Button from '../Button'

import styles from './StatusDisplay.module.scss'

type StatusKind = 'waiting' | 'processing' | 'processed' | 'failed'

type DetailsKind = 'details' | 'help'

type StatusDetails = {
  status: StatusKind
  message?: string
  messageBold?: boolean
  detailsKind?: DetailsKind
  detailsHref?: Url
}

type StatusDisplayProps = {
  children?: React.ReactNode
  className?: string
  coverage: StatusDetails
  checks: StatusDetails
}

interface CellProps extends StatusDetails {
  label: string
}

const orbColor = (status: StatusKind): StatusDotColor => {
  switch (status) {
    case 'waiting':
      return 'grey'
    case 'processing':
      return 'yellow'
    case 'processed':
      return 'green'
    case 'failed':
      return 'red'
  }
}

const StatusCell = ({
  status,
  label,
  message,
  messageBold,
  detailsKind = 'details',
  detailsHref,
}: CellProps) => {
  return (
    <div className={styles.cell}>
      <span className={styles.text}>
        <StatusDot className={styles.dot} color={orbColor(status)} />
        <span className={styles.label}>{label}</span>
        <span
          className={classNames(styles.message, {
            [styles.bold]: messageBold,
          })}
        >
          {message}
        </span>
      </span>
      {detailsHref &&
        (detailsKind == 'details' ? (
          <Button
            className={styles.details}
            href={detailsHref}
            style="inactive"
          >
            Details
          </Button>
        ) : (
          <AccessoryButton
            className={classNames(styles.details, styles.help)}
            href={detailsHref}
            kind="round"
          >
            <HelpIcon />
          </AccessoryButton>
        ))}
    </div>
  )
}

const StatusDisplay = ({ coverage, checks }: StatusDisplayProps) => {
  return (
    <Box className={styles.base}>
      <StatusCell label="Coverage:" {...coverage} />
      <StatusCell label="Checks:" {...checks} />
    </Box>
  )
}

export default StatusDisplay
