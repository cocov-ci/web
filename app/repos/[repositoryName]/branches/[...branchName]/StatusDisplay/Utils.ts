import { StatusDotColor } from 'app/common/StatusDot'
import { HeadProps } from 'types/Commits'

interface StatusDisplayProps {
  message?: string
  statusColor: StatusDotColor
  messageBold: boolean
  detailsHref?: string
}

interface DataProps {
  repositoryName: string
  data: HeadProps
}

export const getCoverage = ({ data, repositoryName }: DataProps) => {
  const { coverage_status, minimum_coverage, coverage_percent, sha } = data

  const coverageObject: StatusDisplayProps = {
    messageBold: false,
  } as StatusDisplayProps

  switch (coverage_status) {
    case 'waiting':
      coverageObject.message = 'Waiting Upload...'
      coverageObject.statusColor = 'grey'
      break
    case 'queued':
    case 'processing':
      coverageObject.message = 'Processing...'
      coverageObject.messageBold = true
      coverageObject.statusColor = 'yellow'
      break
    case 'errored':
      coverageObject.message = 'Error processing'
      coverageObject.messageBold = true
      coverageObject.statusColor = 'red'
      break
    case 'processed':
      if (
        minimum_coverage &&
        coverage_percent &&
        coverage_percent < minimum_coverage
      ) {
        coverageObject.message = `${coverage_percent}% (at least ${minimum_coverage} is required)`
        coverageObject.messageBold = true
        coverageObject.detailsHref = `/repos/${repositoryName}/commits/${sha}/coverage`
        coverageObject.statusColor = 'red'
      } else if (
        !minimum_coverage ||
        (coverage_percent && coverage_percent >= minimum_coverage)
      ) {
        coverageObject.message = `${coverage_percent}%`
        coverageObject.detailsHref = `/repos/${repositoryName}/commits/${sha}/coverage`
        coverageObject.statusColor = 'green'
      }

      break
    default:
  }

  return coverageObject
}

export const getChecks = ({ data, repositoryName }: DataProps) => {
  const { checks_status, issues_count, sha } = data

  const checkObject: StatusDisplayProps = {
    messageBold: false,
  } as StatusDisplayProps

  switch (checks_status) {
    case 'waiting':
    case 'queued':
    case 'processing':
      checkObject.message = 'Processing...'
      checkObject.messageBold = true
      checkObject.detailsHref = `/repos/${repositoryName}/commits/${sha}/checks`
      checkObject.statusColor = 'yellow'
      break
    case 'errored':
      checkObject.message = 'Failed running'
      checkObject.messageBold = true
      checkObject.statusColor = 'red'
      checkObject.detailsHref = `/repos/${repositoryName}/commits/${sha}/checks`
      break
    case 'processed':
      if (issues_count === 0) {
        checkObject.message = 'Passing. No issues detected.'
        checkObject.statusColor = 'green'
      } else {
        checkObject.message = `${issues_count} issues`
        checkObject.detailsHref = `/repos/${repositoryName}/commits/${sha}/issues`
        checkObject.statusColor = 'red'
        checkObject.messageBold = true
      }

      break
    case 'not_configured':
      checkObject.message = 'Not configured.'
      checkObject.messageBold = true
      checkObject.statusColor = 'grey'
      break
    default:
  }

  return checkObject
}
