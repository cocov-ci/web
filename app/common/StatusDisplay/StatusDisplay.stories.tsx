import type { Meta, StoryObj } from '@storybook/react'

import StatusDisplay from './index'

const meta: Meta<typeof StatusDisplay> = {
  title: 'Common/StatusDisplay',
  component: StatusDisplay,
  tags: ['autodocs'],
}

export default meta

type Story = StoryObj<typeof StatusDisplay>

export const Default: Story = {
  args: {
    coverage: {
      status: 'processing',
      messageBold: true,
      message: 'Processing...',
    },
    checks: {
      status: 'processing',
      messageBold: true,
      message: 'Processing...',
    },
  },
}

export const CoverageWaitingChecksProcessing: Story = {
  args: {
    coverage: {
      status: 'waiting',
      message: 'Waiting Upload...',
    },
    checks: {
      status: 'processing',
      messageBold: true,
      message: 'Processing...',
    },
  },
}

export const CoverageProcessedChecksPassing: Story = {
  args: {
    coverage: {
      status: 'processed',
      message: '96.5%',
      detailsHref: '/',
    },
    checks: {
      status: 'processed',
      message: 'Passing. No issues detected.',
    },
  },
}

export const CoverageFailedChecksPassing: Story = {
  args: {
    coverage: {
      status: 'failed',
      messageBold: true,
      message: '89.4% (at least 90% is required)',
      detailsHref: '/',
    },
    checks: {
      status: 'processed',
      message: 'Passing. No issues detected.',
    },
  },
}

export const CoverageErroredChecksPassing: Story = {
  args: {
    coverage: {
      status: 'failed',
      messageBold: true,
      message: 'Error processing.',
      detailsHref: '/',
    },
    checks: {
      status: 'processed',
      message: 'Passing. No issues detected.',
    },
  },
}

export const CoveragePassingChecksFailed: Story = {
  args: {
    coverage: {
      status: 'processed',
      message: '95%',
      detailsHref: '/',
    },
    checks: {
      status: 'failed',
      message: '5 issues',
      messageBold: true,
      detailsHref: '/',
    },
  },
}

export const CoveragePassingChecksErrored: Story = {
  args: {
    coverage: {
      status: 'processed',
      message: '95%',
      detailsHref: '/',
    },
    checks: {
      status: 'failed',
      message: 'Failed running.',
      messageBold: true,
      detailsHref: '/',
    },
  },
}

export const CoveragePassingChecksNotConfigured: Story = {
  args: {
    coverage: {
      status: 'processed',
      message: '95%',
      detailsHref: '/',
    },
    checks: {
      status: 'waiting',
      message: 'Not Configured',
      messageBold: true,
      detailsKind: 'help',
      detailsHref: '/',
    },
  },
}
