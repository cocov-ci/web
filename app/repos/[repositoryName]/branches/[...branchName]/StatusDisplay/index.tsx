import StatusDisplay from 'app/common/StatusDisplay'
import { HeadProps } from 'types/Commits'

import { getChecks, getCoverage } from './Utils'

interface StatusDisplayProps {
  repositoryName: string
  data: HeadProps
}

const StatusDisplayComponent = ({
  data,
  repositoryName,
}: StatusDisplayProps) => {
  if (!data) return null

  return (
    <StatusDisplay
      checks={{ ...getChecks({ data, repositoryName }) }}
      coverage={{ ...getCoverage({ data, repositoryName }) }}
      gutterBottom
    />
  )
}

export default StatusDisplayComponent
