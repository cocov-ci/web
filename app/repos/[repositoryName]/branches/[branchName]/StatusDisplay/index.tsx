import StatusDisplay from 'app/common/StatusDisplay'
import { HeadProps } from 'types/Branches'

import { getChecks, getCoverage } from './Utils'

interface StatusDisplayProps {
  repositoryName: string
  data: HeadProps
}

const StatusDisplayComponent = ({
  data,
  repositoryName,
}: StatusDisplayProps) => {
  return (
    <StatusDisplay
      checks={{ ...getChecks({ data, repositoryName }) }}
      coverage={{ ...getCoverage({ data, repositoryName }) }}
      gutterBottom
    />
  )
}

export default StatusDisplayComponent
