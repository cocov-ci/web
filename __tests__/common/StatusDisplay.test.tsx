import { render, screen } from '@testing-library/react'

import StatusDisplay from 'app/common/StatusDisplay'

import '@testing-library/jest-dom'

const args = {
  coverage: {
    statusColor: 'yellow' as const,
    messageBold: true,
    message: 'Processing...',
  },
  checks: {
    statusColor: 'yellow' as const,
    messageBold: true,
    message: 'Processing...',
  },
}

describe('common/StatusDisplay', () => {
  it('renders StatusDisplay', () => {
    render(<StatusDisplay {...args} />)

    expect(screen.getAllByText('Processing...').length).toEqual(2)
  })

  it('renders StatusDisplay processed', () => {
    const args = {
      coverage: {
        statusColor: 'green' as const,
        message: '96.5%',
        detailsHref: '/',
      },
      checks: {
        statusColor: 'green' as const,
        message: 'Passing. No issues detected.',
      },
    }

    render(<StatusDisplay {...args} />)

    expect(screen.getByText('96.5%')).toBeVisible()
    expect(screen.getByText('Details')).toBeVisible()
    expect(screen.getByText('Passing. No issues detected.')).toBeVisible()
  })

  it('renders StatusDisplay snapshots', () => {
    const { container } = render(<StatusDisplay {...args} />)

    expect(container).toMatchSnapshot()
  })
})
