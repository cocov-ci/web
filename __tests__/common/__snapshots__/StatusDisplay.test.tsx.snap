// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common/StatusDisplay renders StatusDisplay snapshots 1`] = `
<div>
  <div
    class="statusDisplay"
  >
    <div
      class="box base"
    >
      <div
        class="content"
      >
        <div
          class="content"
        >
          <div
            class="cell"
          >
            <span
              class="text"
            >
              <div
                class="dot yellow dot"
              />
              <span
                class="label"
              >
                Coverage:
              </span>
              <span
                class="message bold"
              >
                Processing...
              </span>
            </span>
          </div>
          <div
            class="cell"
          >
            <span
              class="text"
            >
              <div
                class="dot yellow dot"
              />
              <span
                class="label"
              >
                Checks:
              </span>
              <span
                class="message bold"
              >
                Processing...
              </span>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
